Source/WebCore/platform/graphics/ComplexTextController.cpp

```diff
@@ -633,6 +636,7 @@ void ComplexTextController::adjustGlyphsAndAdvances()
     bool runForbidsLeftExpansion = m_run.expansionBehavior().left == ExpansionBehavior::Behavior::Forbid;
     bool runForbidsRightExpansion = m_run.expansionBehavior().right == ExpansionBehavior::Behavior::Forbid;
 
+    UChar previousCharacter = m_textSpacingState.lastCharacterFromPreviousRun;
```

Comment:
Good, but you should also pull out the character's class and cache it, otherwise we're calculating that twice per character and we really only need to do it once.